gobblin-service/src/main/java/org/apache/gobblin/service/modules/utils/FlowCompilationValidationHelper.java

```diff
@@ -140,31 +140,39 @@ public Optional<Dag<JobExecutionPlan>> validateAndHandleConcurrentExecution(Conf
         ConfigurationKeys.FLOW_ALLOW_CONCURRENT_EXECUTION, String.valueOf(this.isFlowConcurrencyEnabled)));
 
     Dag<JobExecutionPlan> jobExecutionPlanDag = specCompiler.compileFlow(flowSpec);
-    if (jobExecutionPlanDag == null || jobExecutionPlanDag.isEmpty()) {
-      return Optional.absent();
-    }
-    addFlowExecutionIdIfAbsent(flowMetadata, jobExecutionPlanDag);
-
-    if (isExecutionPermitted(flowStatusGenerator, flowGroup, flowName, allowConcurrentExecution,
-        Long.parseLong(flowMetadata.get(TimingEvent.FlowEventConstants.FLOW_EXECUTION_ID_FIELD)))) {
-      return Optional.fromNullable(jobExecutionPlanDag);
-    } else {
-      log.warn("Another instance of flowGroup: {}, flowName: {} running; Skipping flow execution since "
-          + "concurrent executions are disabled for this flow.", flowGroup, flowName);
-      sharedFlowMetricsSingleton.conditionallyUpdateFlowGaugeSpecState(flowSpec,
-          SharedFlowMetricsSingleton.CompiledState.SKIPPED);
-      Instrumented.markMeter(sharedFlowMetricsSingleton.getSkippedFlowsMeter());
-      if (!flowSpec.isScheduled()) {
-        // For ad-hoc flow, we might already increase quota, we need to decrease here
-        for (Dag.DagNode dagNode : jobExecutionPlanDag.getStartNodes()) {
-          quotaManager.releaseQuota(dagNode);
+    Optional<Dag<JobExecutionPlan>> optionalJobExecutionPlan = null;
+
+    try {
+      if (jobExecutionPlanDag == null || jobExecutionPlanDag.isEmpty()) {
+        optionalJobExecutionPlan = Optional.absent();
+      }
+      addFlowExecutionIdIfAbsent(flowMetadata, jobExecutionPlanDag);
+
+      if (isExecutionPermitted(flowStatusGenerator, flowGroup, flowName, allowConcurrentExecution,
+          Long.parseLong(flowMetadata.get(TimingEvent.FlowEventConstants.FLOW_EXECUTION_ID_FIELD)))) {
+        optionalJobExecutionPlan = Optional.fromNullable(jobExecutionPlanDag);
+      } else {
+        log.warn("Another instance of flowGroup: {}, flowName: {} running; Skipping flow execution since "
+            + "concurrent executions are disabled for this flow.", flowGroup, flowName);
+        optionalJobExecutionPlan = Optional.absent();
+      }
+    } finally {
+      if (optionalJobExecutionPlan == null || !optionalJobExecutionPlan.isPresent()) {
```

Comment:
again, convoluted to need to handle a case where a var may not yet have been initialized (even ironic, since the point of `Optional` is that we no longer need `null` checks)

where are the test cases to validate correctness?

Reply:
updated this code path to make it more clear. I realized the quota manager does not need to be updated when the compilation fails so this looks much simpler now